File: sample/OpenCvSharp.Samples/Samples/FlannTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using OpenCvSharp;
using OpenCvSharp.CPlusPlus;
using OpenCvSharp.CPlusPlus.Flann;

namespace OpenCvSharp.Test
{
    /// <summary>
    /// cv::flann
    /// </summary>
    class FlannTest
    {
        public FlannTest()
        {
            Console.WriteLine("===== FlannTest =====");

            // creates data set
            using (Mat features = new Mat(10000, 2, MatrixType.F32C1))
            {
                Random rand = new Random();
                for (int i = 0; i < features.Rows; i++)
                {
                    features.Set<float>(i, 0, rand.Next(10000));
                    features.Set<float>(i, 1, rand.Next(10000));
                }

                // query
                CvPoint2D32f queryPoint = new CvPoint2D32f(7777, 7777);
                Mat queries = new Mat(1, 2, MatrixType.F32C1);
                queries.Set<float>(0, 0, queryPoint.X);
                queries.Set<float>(0, 1, queryPoint.Y);
                Console.WriteLine("query:({0}, {1})", queryPoint.X, queryPoint.Y);
                Console.WriteLine("-----");

                // knnSearch
                using (Index nnIndex = new Index(features, new KDTreeIndexParams(4)))
                {
                    int knn = 1;
                    int[] indices;
                    float[] dists;
                    nnIndex.KnnSearch(queries, out indices, out dists, knn, new SearchParams(32));

                    for (int i = 0; i < knn; i++)
                    {
                        int index = indices[i];
                        float dist = dists[i];
                        CvPoint2D32f pt = new CvPoint2D32f(features.Get<float>(index, 0), features.Get<float>(index, 1));
                        Console.Write("No.{0}\t", i);
                        Console.Write("index:{0}", index);
                        Console.Write(" distance:{0}", dist);
                        Console.Write(" data:({0}, {1})", pt.X, pt.Y);
                        Console.WriteLine();
                    }
                    Console.Read();
                }
            }
        }
    }
}

File: sample/CStyleSamplesVB/Samples/FlannTest.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports OpenCvSharp
Imports OpenCvSharp.CPlusPlus

' Namespace OpenCvSharpSamplesVB
''' <summary>
''' cv::flann
''' </summary>
Friend Module FlannTest
    Public Sub Start()
        Form1.TextBox1.AppendText(Environment.NewLine & String.Format(("===== FlannTest =====")))

        ' creates data set
        Using features As New Mat(10000, 2, MatrixType.F32C1)
            Dim rand As New Random()
            For i As Integer = 0 To features.Rows - 1
                features.Set(Of Single)(i, 0, rand.Next(10000))
                features.Set(Of Single)(i, 1, rand.Next(10000))
            Next i

            ' query
            Dim queryPoint As New CvPoint2D32f(7777, 7777)
            Dim queries As New Mat(1, 2, MatrixType.F32C1)
            queries.Set(Of Single)(0, 0, queryPoint.X)
            queries.Set(Of Single)(0, 1, queryPoint.Y)
            Form1.TextBox1.AppendText(Environment.NewLine & String.Format("query:({0}, {1})", queryPoint.X, queryPoint.Y))
            Form1.TextBox1.AppendText(Environment.NewLine & String.Format("-----"))

            ' knnSearch
            Using nnIndex As New Flann.Index(features, New Flann.KDTreeIndexParams(4))
                Dim knn As Integer = 1
                Dim indices() As Integer
                Dim dists() As Single
                nnIndex.KnnSearch(queries, indices, dists, knn, New Flann.SearchParams(32))

                For i As Integer = 0 To knn - 1
                    Dim index As Integer = indices(i)
                    Dim dist As Single = dists(i)
                    Dim pt As New CvPoint2D32f(features.Get(Of Single)(index, 0), features.Get(Of Single)(index, 1))
                    Form1.TextBox1.AppendText(Environment.NewLine & String.Format("No.{0}" & vbTab, i))
                    Form1.TextBox1.AppendText(Environment.NewLine & String.Format("index:{0}", index))
                    Form1.TextBox1.AppendText(Environment.NewLine & String.Format(" distance:{0}", dist))
                    Form1.TextBox1.AppendText(Environment.NewLine & String.Format(" data:({0}, {1})", pt.X, pt.Y))
                    Form1.TextBox1.AppendText(Environment.NewLine & " ")
                Next i
                'Console.Read()
            End Using
        End Using
    End Sub
End Module
' End Namespace
